"use strict";(self.webpackChunkdocumentation=self.webpackChunkdocumentation||[]).push([[2790],{8772:(e,i,t)=>{t.r(i),t.d(i,{assets:()=>l,contentTitle:()=>r,default:()=>c,frontMatter:()=>s,metadata:()=>a,toc:()=>h});var n=t(5893),o=t(1151);const s={sidebar_position:7},r="Memory profiling",a={id:"ChromOptimise/Memory-Profiling",title:"Memory profiling",description:"To gain insights into the peak heap consumption of each script in the pipeline, we utilized the tool Heaptrack. This allowed for the estimation of memory consumption based on unit tests (albeit only with relatively small amounts of data). Utilizing this information will empower the user to make well-informed decisions regarding the allocation of memory for the scripts involved in the pipeline and ChromHMM.",source:"@site/docs/ChromOptimise/Memory-Profiling.md",sourceDirName:"ChromOptimise",slug:"/ChromOptimise/Memory-Profiling",permalink:"/ChromOptimise/ChromOptimise/Memory-Profiling",draft:!1,unlisted:!1,tags:[],version:"current",sidebarPosition:7,frontMatter:{sidebar_position:7},sidebar:"documentationSidebar",previous:{title:"Processing times",permalink:"/ChromOptimise/ChromOptimise/Processing-Times"},next:{title:"SLURM information",permalink:"/ChromOptimise/ChromOptimise/SLURM-Workload-Manager-Information"}},l={},h=[{value:"0_DownloadBluePrint.sh",id:"0_downloadblueprintsh",level:2},{value:"1_MoveFilesToSingleDirectory.sh",id:"1_movefilestosingledirectorysh",level:2},{value:"2_batch_ProcessBamFiles.sh",id:"2_batch_processbamfilessh",level:2},{value:"3_SubsampleBamFiles.sh",id:"3_subsamplebamfilessh",level:2},{value:"4_BinarizeBamFiles.sh",id:"4_binarizebamfilessh",level:2},{value:"5_batch_CreateIncrementalModels.sh",id:"5_batch_createincrementalmodelssh",level:2},{value:"6_OptimalNumberofStates.sh",id:"6_optimalnumberofstatessh",level:2},{value:"7_ReferenceLDSCore.sh",id:"7_referenceldscoresh",level:2},{value:"8_PartitionedHeritability.sh",id:"8_partitionedheritabilitysh",level:2},{value:"Generate_Big_Model.sh",id:"generate_big_modelsh",level:2}];function m(e){const i={a:"a",br:"br",code:"code",h1:"h1",h2:"h2",li:"li",p:"p",ul:"ul",...(0,o.a)(),...e.components};return(0,n.jsxs)(n.Fragment,{children:[(0,n.jsx)(i.h1,{id:"memory-profiling",children:"Memory profiling"}),"\n",(0,n.jsxs)(i.p,{children:["To gain insights into the peak heap consumption of each script in the pipeline, we utilized the tool Heaptrack. This allowed for the estimation of memory consumption based on unit tests (albeit only with relatively small amounts of data). Utilizing this information will empower the user to make well-informed decisions regarding the allocation of memory for the scripts involved in the pipeline and ChromHMM.\n",(0,n.jsx)(i.br,{}),"\n",(0,n.jsx)(i.br,{}),"\n","Heaptrack, produced by KDE, can be obtained through an appimage download or local compilation. The GitHub repository for Heaptrack can be accessed ",(0,n.jsx)(i.a,{href:"https://github.com/KDE/heaptrack",children:"here"}),"."]}),"\n",(0,n.jsx)(i.h2,{id:"0_downloadblueprintsh",children:"0_DownloadBluePrint.sh"}),"\n",(0,n.jsx)(i.p,{children:"The memory consumption for this script depends on the abundance and size of the files that are being downloaded from EGA repositories."}),"\n",(0,n.jsx)(i.h2,{id:"1_movefilestosingledirectorysh",children:"1_MoveFilesToSingleDirectory.sh"}),"\n",(0,n.jsxs)(i.p,{children:["The memory usage for this script is typically minimal, thanks to the way Linux manages file movements without transferring data in memory, but rather updating the pointer in the file management system. Consequently, even when dealing with large multi-gigabyte files, the peak memory consumption remains in the kilobyte range. Hence, this script requires very little memory.\n",(0,n.jsx)(i.br,{}),"\n",(0,n.jsx)(i.br,{}),"\n","However, if there is an attempt to move files between different drives, the above characteristics no longer hold true, and a much higher peak memory consumption is expected due to the need for redundancy plans to address potential corruption."]}),"\n",(0,n.jsx)(i.h2,{id:"2_batch_processbamfilessh",children:"2_batch_ProcessBamFiles.sh"}),"\n",(0,n.jsxs)(i.p,{children:["The primary source of memory consumption in this script is the execution of the ",(0,n.jsx)(i.code,{children:"samtools sort"})," command. This operation requires significantly more memory than any other part of the script due to the large number of temporary allocations it performs.\n",(0,n.jsx)(i.br,{}),"\n","In testing, the following was observed:"]}),"\n",(0,n.jsxs)(i.ul,{children:["\n",(0,n.jsxs)(i.li,{children:["For a 1.5 GB compressed .bam file:","\n",(0,n.jsxs)(i.ul,{children:["\n",(0,n.jsxs)(i.li,{children:[(0,n.jsx)(i.code,{children:"samtools index"})," had 22.2 MB peak memory consumption"]}),"\n",(0,n.jsxs)(i.li,{children:[(0,n.jsx)(i.code,{children:"samtools idxstats"})," had 2.6 MB peak memory consumption"]}),"\n",(0,n.jsxs)(i.li,{children:[(0,n.jsx)(i.code,{children:"samtools stats"})," had 5 MB peak memory consumption"]}),"\n",(0,n.jsxs)(i.li,{children:[(0,n.jsx)(i.code,{children:"samtools view"})," had 4.5 MB peak memory consumption"]}),"\n",(0,n.jsxs)(i.li,{children:[(0,n.jsx)(i.code,{children:"samtools sort"})," had 927 MB peak memory consumption"]}),"\n"]}),"\n"]}),"\n"]}),"\n",(0,n.jsxs)(i.p,{children:["Based on this information, it can be inferred that the peak memory consumption will correspond to the size of the largest .bam file processed by the script, with the peak consumption expected to be slightly less than the size of the largest file.\n",(0,n.jsx)(i.br,{}),"\n","Additionally, it's important to note that this script is designed to be executed as an array through the SLURM workload manager. Therefore, the peak memory consumption is anticipated to be proportional to the size of the largest .bam files multiplied by the number of array elements."]}),"\n",(0,n.jsx)(i.h2,{id:"3_subsamplebamfilessh",children:"3_SubsampleBamFiles.sh"}),"\n",(0,n.jsxs)(i.p,{children:["This script is designed to utilize minimal memory when merging and subsampling a small number of files. During the merging process, ",(0,n.jsx)(i.code,{children:"samtools merge"})," only opens a small component of each file being merged at a time, instead of opening each file in its entirety. The subsampling is carried out using ",(0,n.jsx)(i.code,{children:"samtools view"}),", which uses very little peak memory consumption relative to file size (see ",(0,n.jsx)(i.a,{href:"#2_batch_processbamfilessh",children:"2_batch_ProcessBamFiles.sh"}),").\n",(0,n.jsx)(i.br,{}),"\n","However, challenges may arise when merging a significantly larger number of files. According to the ",(0,n.jsx)(i.a,{href:"http://www.htslib.org/doc/samtools-merge.html",children:"documentation"})," for ",(0,n.jsx)(i.code,{children:"samtools merge"}),", the command utilizes ~1 MB of data from each file at a time. It's important to be mindful of this information when merging large numbers of files within the pipeline. In the event of memory-related errors with this script, it is recommended to execute ",(0,n.jsx)(i.code,{children:"ulimit -n"})," in the terminal to check the limit set on how many files can be open at one time. If the number of files to merge exceeds the output of ",(0,n.jsx)(i.code,{children:"ulimit -n"}),", it is advisable to consider modifying the script to merge the files in batches instead of all at once."]}),"\n",(0,n.jsx)(i.h2,{id:"4_binarizebamfilessh",children:"4_BinarizeBamFiles.sh"}),"\n",(0,n.jsxs)(i.p,{children:["ChromHMM's ",(0,n.jsx)(i.code,{children:"BinarizeBam"})," command is the main contributor to memory consumption in this script.\n",(0,n.jsx)(i.br,{}),"\n","In testing, the following was observed:"]}),"\n",(0,n.jsxs)(i.ul,{children:["\n",(0,n.jsx)(i.li,{children:"For a directory with 16.8 GB of .bam files: peak heap memory consumption was 40 MB"}),"\n"]}),"\n",(0,n.jsx)(i.h2,{id:"5_batch_createincrementalmodelssh",children:"5_batch_CreateIncrementalModels.sh"}),"\n",(0,n.jsxs)(i.p,{children:["ChromHMM's ",(0,n.jsx)(i.code,{children:"LearnModel"})," command is the main contributor to memory consumption in this script.\n",(0,n.jsx)(i.br,{}),"\n","In testing, the following was observed:"]}),"\n",(0,n.jsxs)(i.ul,{children:["\n",(0,n.jsxs)(i.li,{children:["Using binary files that were 1 MB in total:","\n",(0,n.jsxs)(i.ul,{children:["\n",(0,n.jsx)(i.li,{children:"2 state model training had 181 MB peak memory consumption"}),"\n",(0,n.jsx)(i.li,{children:"3 state model training had 145 MB peak memory consumption"}),"\n",(0,n.jsx)(i.li,{children:"4 state model training had 165 MB peak memory consumption"}),"\n",(0,n.jsx)(i.li,{children:"5 state model training had 128 MB peak memory consumption"}),"\n",(0,n.jsx)(i.li,{children:"6 state model training had 165 MB peak memory consumption"}),"\n",(0,n.jsx)(i.li,{children:"7 state model training had 168 MB peak memory consumption"}),"\n"]}),"\n"]}),"\n"]}),"\n",(0,n.jsx)(i.p,{children:"Based on the provided information, it is evident that the memory consumption is primarily dependent on the total size of the binary files rather than the number of states used in the models.\n/\nAdditionally, it's important to note that this script is designed to be executed as an array through the SLURM workload manager. Therefore, the peak memory consumption is anticipated to be proportional to the size of the largest .bam files multiplied by the number of array elements."}),"\n",(0,n.jsx)(i.h2,{id:"6_optimalnumberofstatessh",children:"6_OptimalNumberofStates.sh"}),"\n",(0,n.jsx)(i.p,{children:"This script utilizes Rscripts which entail minor computations, resulting in very low memory consumption. Through testing, it has been observed that the memory usage never surpassed a few kilobytes."}),"\n",(0,n.jsx)(i.h2,{id:"7_referenceldscoresh",children:"7_ReferenceLDSCore.sh"}),"\n",(0,n.jsx)(i.p,{children:"This script's biggest requirement of memory comes from the SNP assigment R script\nthat it calls. In testing, chromosomes 1 and 2 use about 800MB of memory.\nBecause the script is ran as an array that processes each chromosome at the same\ntime, 22 GB of memory is allocated."}),"\n",(0,n.jsx)(i.h2,{id:"8_partitionedheritabilitysh",children:"8_PartitionedHeritability.sh"}),"\n",(0,n.jsxs)(i.p,{children:["According to ",(0,n.jsx)(i.a,{href:"https://github.com/bulik/ldsc/wiki/FAQ",children:"ldsc's creators"}),",\npartitioned heritability should only take up ~8GB for 50 annotations. The number\nof annotations in this script will be greater than 50. In testing, over 20GB is\nrequired to stop out of memory errors for just 58 annotations. To account for\nthis, the SLURM directive is set to 50GB of memory."]}),"\n",(0,n.jsx)(i.h2,{id:"generate_big_modelsh",children:"Generate_Big_Model.sh"}),"\n",(0,n.jsxs)(i.p,{children:["This script is similar to ",(0,n.jsx)(i.a,{href:"#5_batch_createincrementalmodelssh",children:"5_batch_CreateIncrementalModels.sh"}),", suggesting that the memory consumption pattern is likely to align with the conclusions drawn there. However, it's important to note that due to the computational time required for this script, utilizing heaptrack to monitor memory allocations was impractical. Consequently, the memory consumption for notably large models (20 states or more) remains unknown and is not extensively documented."]})]})}function c(e={}){const{wrapper:i}={...(0,o.a)(),...e.components};return i?(0,n.jsx)(i,{...e,children:(0,n.jsx)(m,{...e})}):m(e)}},1151:(e,i,t)=>{t.d(i,{Z:()=>a,a:()=>r});var n=t(7294);const o={},s=n.createContext(o);function r(e){const i=n.useContext(s);return n.useMemo((function(){return"function"==typeof e?e(i):{...i,...e}}),[i,e])}function a(e){let i;return i=e.disableParentContext?"function"==typeof e.components?e.components(o):e.components||o:r(e.components),n.createElement(s.Provider,{value:i},e.children)}}}]);