"use strict";(self.webpackChunkdocumentation=self.webpackChunkdocumentation||[]).push([[9929],{9049:(e,n,t)=>{t.r(n),t.d(n,{assets:()=>a,contentTitle:()=>s,default:()=>h,frontMatter:()=>r,metadata:()=>l,toc:()=>c});var i=t(5893),o=t(1151);const r={title:"4_ReferenceLDSCore",description:"Calculates the ldscores for annotation files.",sidebar_position:1},s="4_ReferenceLDSCore",l={id:"ChromOptimise/Pipeline-Explanation/LDSC/ReferenceLDSCore",title:"4_ReferenceLDSCore",description:"Calculates the ldscores for annotation files.",source:"@site/docs/ChromOptimise/Pipeline-Explanation/LDSC/4_ReferenceLDSCore.md",sourceDirName:"ChromOptimise/Pipeline-Explanation/LDSC",slug:"/ChromOptimise/Pipeline-Explanation/LDSC/ReferenceLDSCore",permalink:"/ChromOptimise/ChromOptimise/Pipeline-Explanation/LDSC/ReferenceLDSCore",draft:!1,unlisted:!1,tags:[],version:"current",sidebarPosition:1,frontMatter:{title:"4_ReferenceLDSCore",description:"Calculates the ldscores for annotation files.",sidebar_position:1},sidebar:"documentationSidebar",previous:{title:"LDSC integration",permalink:"/ChromOptimise/category/ldsc-integration"},next:{title:"SNPAssignment",permalink:"/ChromOptimise/ChromOptimise/Pipeline-Explanation/LDSC/SNPAssignment"}},a={},c=[{value:"Explanation",id:"explanation",level:2}];function p(e){const n={a:"a",h1:"h1",h2:"h2",li:"li",p:"p",ul:"ul",...(0,o.a)(),...e.components};return(0,i.jsxs)(i.Fragment,{children:[(0,i.jsx)(n.h1,{id:"4_referenceldscore",children:"4_ReferenceLDSCore"}),"\n",(0,i.jsx)(n.h2,{id:"explanation",children:"Explanation"}),"\n",(0,i.jsxs)(n.p,{children:["This is a script that will run 2 R scripts and also\n",(0,i.jsx)(n.a,{href:"https://github.com/bulik/ldsc",children:"LDSC"}),", which outputs information useful for\ndetermining if a ChromHMM model has biologically relevant states (not just\nstatistically relevant)."]}),"\n",(0,i.jsxs)(n.p,{children:["The script achieves this using\n",(0,i.jsx)(n.a,{href:"https://www.nature.com/articles/ng.3404",children:"partitioned heritability"})," and\nrequires reference files in order to work properly. These files can be\ndownloaded from ",(0,i.jsx)(n.a,{href:"https://zenodo.org/records/10515792",children:"this online repository"}),".\nThe reference files do not have to be from the 1000 genomes project (just any\nlarge collection of SNPs will do). The required files are:"]}),"\n",(0,i.jsxs)(n.ul,{children:["\n",(0,i.jsx)(n.li,{children:"1000 genomes (or similar) PLINK files (for your genome build)"}),"\n",(0,i.jsx)(n.li,{children:"1000 genomes (or similar) weight files (built from ldsc, best to get these\nfrom the online repository)"}),"\n",(0,i.jsx)(n.li,{children:"Collection of GWAS traits in sumstats format (again, best to get these from\nthe online repository)"}),"\n"]})]})}function h(e={}){const{wrapper:n}={...(0,o.a)(),...e.components};return n?(0,i.jsx)(n,{...e,children:(0,i.jsx)(p,{...e})}):p(e)}},1151:(e,n,t)=>{t.d(n,{Z:()=>l,a:()=>s});var i=t(7294);const o={},r=i.createContext(o);function s(e){const n=i.useContext(r);return i.useMemo((function(){return"function"==typeof e?e(n):{...n,...e}}),[n,e])}function l(e){let n;return n=e.disableParentContext?"function"==typeof e.components?e.components(o):e.components||o:s(e.components),i.createElement(r.Provider,{value:n},e.children)}}}]);