<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper plugin-docs plugin-id-default docs-version-current docs-doc-page docs-doc-id-ChromOptimise/Pipeline-Explanation" data-has-hydrated="false">
<head>
<meta charset="UTF-8">
<meta name="generator" content="Docusaurus v3.1.0">
<title data-rh="true">Pipeline explanation | ChromOptimise</title><meta data-rh="true" name="viewport" content="width=device-width,initial-scale=1"><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:image" content="https://sof202.github.io/ChromOptimise/img/dna.png"><meta data-rh="true" name="twitter:image" content="https://sof202.github.io/ChromOptimise/img/dna.png"><meta data-rh="true" property="og:url" content="https://sof202.github.io/ChromOptimise/ChromOptimise/Pipeline-Explanation"><meta data-rh="true" property="og:locale" content="en"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="Pipeline explanation | ChromOptimise"><meta data-rh="true" name="description" content="The main pipeline is split into 3 main steps:"><meta data-rh="true" property="og:description" content="The main pipeline is split into 3 main steps:"><link data-rh="true" rel="icon" href="/ChromOptimise/img/dna.ico"><link data-rh="true" rel="canonical" href="https://sof202.github.io/ChromOptimise/ChromOptimise/Pipeline-Explanation"><link data-rh="true" rel="alternate" href="https://sof202.github.io/ChromOptimise/ChromOptimise/Pipeline-Explanation" hreflang="en"><link data-rh="true" rel="alternate" href="https://sof202.github.io/ChromOptimise/ChromOptimise/Pipeline-Explanation" hreflang="x-default"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.13.24/dist/katex.min.css" integrity="sha384-odtC+0UGzzFL/6PNoE8rX/SPcQDXBJ+uRepguP4QkPCm2LBxH3FA3y+fKSiJ+AmM" crossorigin="anonymous"><link rel="stylesheet" href="/ChromOptimise/assets/css/styles.e3ab61ce.css">
<script src="/ChromOptimise/assets/js/runtime~main.e32e51d0.js" defer="defer"></script>
<script src="/ChromOptimise/assets/js/main.62103c36.js" defer="defer"></script>
</head>
<body class="navigation-with-keyboard">
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){try{return new URLSearchParams(window.location.search).get("docusaurus-theme")}catch(t){}}()||function(){try{return localStorage.getItem("theme")}catch(t){}}();t(null!==e?e:"light")}(),function(){try{const c=new URLSearchParams(window.location.search).entries();for(var[t,e]of c)if(t.startsWith("docusaurus-data-")){var a=t.replace("docusaurus-data-","data-");document.documentElement.setAttribute(a,e)}}catch(t){}}()</script><div id="__docusaurus"><div role="region" aria-label="Skip to main content"><a class="skipToContent_fXgn" href="#__docusaurus_skipToContent_fallback">Skip to main content</a></div><nav aria-label="Main" class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Toggle navigation bar" aria-expanded="false" class="navbar__toggle clean-btn" type="button"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/ChromOptimise/"><div class="navbar__logo"><img src="/ChromOptimise/img/logo.svg" alt="Site Logo" class="themedComponent_mlkZ themedComponent--light_NVdE"><img src="/ChromOptimise/img/logo.svg" alt="Site Logo" class="themedComponent_mlkZ themedComponent--dark_xIcU"></div><b class="navbar__title text--truncate">ChromOptimise</b></a><a aria-current="page" class="navbar__item navbar__link navbar__link--active" href="/ChromOptimise/">Documentation</a></div><div class="navbar__items navbar__items--right"><a href="https://github.com/sof202/ChromOptimise" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_nPIU"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a><div class="toggle_vylO colorModeToggle_DEke"><button class="clean-btn toggleButton_gllP toggleButtonDisabled_aARS" type="button" disabled="" title="Switch between dark and light mode (currently light mode)" aria-label="Switch between dark and light mode (currently light mode)" aria-live="polite"><svg viewBox="0 0 24 24" width="24" height="24" class="lightToggleIcon_pyhR"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" class="darkToggleIcon_wfgR"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg></button></div><div class="navbarSearchContainer_Bca1"></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div id="__docusaurus_skipToContent_fallback" class="main-wrapper mainWrapper_z2l0"><div class="docsWrapper_hBAB"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_sjWU" type="button"></button><div class="docRoot_UBD9"><aside class="theme-doc-sidebar-container docSidebarContainer_YfHR"><div class="sidebarViewport_aRkj"><div class="sidebar_njMd"><nav aria-label="Docs sidebar" class="menu thin-scrollbar menu_SIkG"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/ChromOptimise/">Welcome to the ChromOptimise wiki!</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--active" aria-expanded="true" href="/ChromOptimise/category/chromoptimise-documentation">ChromOptimise Documentation</a><button aria-label="Collapse sidebar category &#x27;ChromOptimise Documentation&#x27;" type="button" class="clean-btn menu__caret"></button></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/ChromOptimise/ChromOptimise/ChromHMM-overview">ChromHMM overview</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/ChromOptimise/ChromOptimise/Configuration-Files-Setup">Configuration files setup</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/ChromOptimise/ChromOptimise/Pipeline-Explanation">Pipeline explanation</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/ChromOptimise/ChromOptimise/Supplementary-pipeline-explanation">Supplementary pipeline explanation</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/ChromOptimise/ChromOptimise/Factors-that-affect-the-output">Factors that affect the output</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/ChromOptimise/ChromOptimise/Processing-Times">Processing times</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/ChromOptimise/ChromOptimise/Memory-Profiling">Memory profiling</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/ChromOptimise/ChromOptimise/SLURM-Workload-Manager-Information">SLURM information</a></li></ul></li></ul></nav></div></div></aside><main class="docMainContainer_TBSr"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_VOVn"><div class="docItemContainer_Djhp"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Z_bl" aria-label="Breadcrumbs"><ul class="breadcrumbs" itemscope="" itemtype="https://schema.org/BreadcrumbList"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/ChromOptimise/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_YNFT"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item"><a class="breadcrumbs__link" itemprop="item" href="/ChromOptimise/category/chromoptimise-documentation"><span itemprop="name">ChromOptimise Documentation</span></a><meta itemprop="position" content="1"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link" itemprop="name">Pipeline explanation</span><meta itemprop="position" content="2"></li></ul></nav><div class="tocCollapsible_ETCw theme-doc-toc-mobile tocMobile_ITEo"><button type="button" class="clean-btn tocCollapsibleButton_TO0P">On this page</button></div><div class="theme-doc-markdown markdown"><h1>Pipeline explanation</h1>
<p>The main pipeline is split into 3 main steps:</p>
<ul>
<li>Manipulate the data set using user defined parameters</li>
<li>Use ChromHMM&#x27;s pipeline to obtain multiple hidden Markov models on the processed data</li>
<li>Use statistical analysis to evaluate the model with the optimum number of states</li>
</ul>
<p><br>
<!-- -->The pipeline expects aligned <code>.bam</code> files as its input. One may want to download these files from EGA, this can be done using <a href="#0_egadownloadingsh">0_EGADownloading.sh</a>.
<br>
<!-- -->The pipeline will then output:</p>
<ul>
<li>A scatter plot that showcases how the number of states in the hidden Markov model relates to the estimated log likelihood of the model</li>
<li>A file containing the optimum number of states to use with the binarized data files (when a specific bin size has been used)</li>
</ul>
<p>For a visual representation of the pipeline, please consult the file: <a href="https://github.com/sof202/ChromOptimise/blob/main/information/Optimal_States_Analysis_Pipeline.pdf" target="_blank" rel="noopener noreferrer"><code>Optimal_States_Analysis_Pipeline.pdf</code></a>.</p>
<h2 class="anchor anchorWithStickyNavbar_LWe7" id="0_egadownloadingsh">0_EGADownloading.sh<a href="#0_egadownloadingsh" class="hash-link" aria-label="Direct link to 0_EGADownloading.sh" title="Direct link to 0_EGADownloading.sh">​</a></h2>
<p>This is an artefact of the pipeline originally being built for the blueprint data from EGA.
The script will download files from a file of file names using the pyega3 python package.
There are a few prerequisites to use this properly:</p>
<p>Step one: Login credentials need to be saved in <code>.json</code> format:</p>
<div class="language-json codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-json codeBlock_bY9V thin-scrollbar" style="color:#393A34;background-color:#f6f8fa"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">{</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    &quot;username&quot;:&quot;username@domain.com&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    &quot;password&quot;:&quot;password&quot;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div>
<p>Step Two: Create a conda environment with pyega3 installed:</p>
<div class="language-shell codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-shell codeBlock_bY9V thin-scrollbar" style="color:#393A34;background-color:#f6f8fa"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">conda create -n myenv pyega3</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div>
<p>Step Three: Change the following paths:</p>
<ul>
<li>Path to file of file names</li>
<li>Path to conda config file (<code>.../etc/profile.d/conda.sh</code>)</li>
<li>Path to login credentials created in step 1</li>
<li>Path to conda environment created in step 2</li>
</ul>
<p>Example:</p>
<div class="language-shell codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-shell codeBlock_bY9V thin-scrollbar" style="color:#393A34;background-color:#f6f8fa"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># Downloads all files listed in FileOfFileNames.txt from EGA</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sbatch 0_EGADownloading.sh FileOfFileNames.txt</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div>
<h2 class="anchor anchorWithStickyNavbar_LWe7" id="1_movefilestosingledirectorysh">1_MoveFilesToSingleDirectory.sh<a href="#1_movefilestosingledirectorysh" class="hash-link" aria-label="Direct link to 1_MoveFilesToSingleDirectory.sh" title="Direct link to 1_MoveFilesToSingleDirectory.sh">​</a></h2>
<p>Again, this is an artefact of the pipeline originally being built for the blueprint data from EGA.</p>
<p>The downloaded files from EGA will be in separate directories (as each directory has its own checksum file to validate the downloads). For easier file manipulation later in the pipeline, these files are collated into folders denoted by the epigenetic mark they contain information on.</p>
<div class="theme-admonition theme-admonition-warning admonition_xJq3 alert alert--warning"><div class="admonitionHeading_Gvgb"><span class="admonitionIcon_Rf37"><svg viewBox="0 0 16 16"><path fill-rule="evenodd" d="M8.893 1.5c-.183-.31-.52-.5-.887-.5s-.703.19-.886.5L.138 13.499a.98.98 0 0 0 0 1.001c.193.31.53.501.886.501h13.964c.367 0 .704-.19.877-.5a1.03 1.03 0 0 0 .01-1.002L8.893 1.5zm.133 11.497H6.987v-2.003h2.039v2.003zm0-3.004H6.987V5.987h2.039v4.006z"></path></svg></span>Important</div><div class="admonitionContent_BuS1"><p>This script relies on the <code>.bam</code> files containing the epigenetic mark in the file names. Please ensure this is the case for your raw files.</p></div></div>
<p>Example:</p>
<div class="language-shell codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-shell codeBlock_bY9V thin-scrollbar" style="color:#393A34;background-color:#f6f8fa"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># Moves all files containing &#x27;H3K27me3&#x27; to: ${RAW_DIR}/H3k27me3</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sbatch 1_MoveFilesToSingleDirectory.sh H3K27me3</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div>
<p><strong>Note</strong>: If this step is skipped, please ensure that the raw <code>.bam</code> files are organised into folders with memorable names (epigenetic mark name is reccomended).</p>
<h2 class="anchor anchorWithStickyNavbar_LWe7" id="2_batch_processbamfilessh">2_batch_ProcessBamFiles.sh<a href="#2_batch_processbamfilessh" class="hash-link" aria-label="Direct link to 2_batch_ProcessBamFiles.sh" title="Direct link to 2_batch_ProcessBamFiles.sh">​</a></h2>
<p>If the aligned <code>.bam</code> files you are working with have already been processed, skip this step.</p>
<p>The purpose of this step is to sort, index, filter and remove the duplicates of the aligned <code>.bam</code> files.
<br>
<!-- -->Note that, in the blueprint data, the aligned read files are single end and duplicates have already been marked for duplicate reads using picard.
<br>
<!-- -->Therefore, the processing stage will need to be adapted if your <code>.bam</code> files don&#x27;t have these properties.</p>
<p>This step also outputs some statistics on the aligned read files using <code>samtools stats</code> and <code>samtools idxstats</code> before and after processing.</p>
<p>Example:</p>
<div class="language-shell codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-shell codeBlock_bY9V thin-scrollbar" style="color:#393A34;background-color:#f6f8fa"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># Processes all files in ${RAW_DIR}/H3k27me3 using a Phred score threshold</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># of 30</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sbatch --array=1-4 2_batch_ProcessBamFiles.sh H3K27me3 30</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div>
<h2 class="anchor anchorWithStickyNavbar_LWe7" id="3_subsamplebamfilessh">3_SubsampleBamFiles.sh<a href="#3_subsamplebamfilessh" class="hash-link" aria-label="Direct link to 3_SubsampleBamFiles.sh" title="Direct link to 3_SubsampleBamFiles.sh">​</a></h2>
<p>This merges all of the processed <code>.bam</code> files from the previous step using <code>samtools merge</code> and subsequently subsamples this file using <code>samtools view -s</code>.
<br>
<strong>Note</strong>: samtools had a known bug in version 0.1.18 where subsampling would fail if the sample size was above 50%, ensure that your version of samtools satisfies the <a href="/ChromOptimise/#software-requirements">software requirements</a>.</p>
<p>The reason for merging files and subsampling instead of simply subsampling the original files is that it is unlikely that every processed <code>.bam</code> file is of the same size. If the <code>.bam</code> files are of different sizes, then different subsamples with the same number of files can have different sizes.
<br>
<!-- -->This results in the different samples no longer being directly comparable.</p>
<p><strong>Note</strong>: If you are working with a lot of files, this step can use a large amount of memory due to the way samtools merges files, see <a href="/ChromOptimise/ChromOptimise/Memory-Profiling">memory profiling</a> for more information.</p>
<p>Example:</p>
<div class="language-shell codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-shell codeBlock_bY9V thin-scrollbar" style="color:#393A34;background-color:#f6f8fa"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># Merges all files in ${PROCESSED_DIR}/H3k27me3 and subsamples this file</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># with sample size 75%</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sbatch 3_SubsampleBamFiles.sh H3K27me3 75</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div>
<h2 class="anchor anchorWithStickyNavbar_LWe7" id="4_binarizefilessh">4_BinarizeFiles.sh<a href="#4_binarizefilessh" class="hash-link" aria-label="Direct link to 4_BinarizeFiles.sh" title="Direct link to 4_BinarizeFiles.sh">​</a></h2>
<p>This uses ChromHMM&#x27;s <code>BinarizeBam</code> command to binarize the subsampled <code>.bam</code> file produced in the previous step.
<br>
<!-- -->It is assumed at this point that the above steps have been completed for all of the epigenetic marks that you aim to use in the hidden Markov models, and that the sample size used for each mark is the same.
<br>
<!-- -->This step takes a user input for the bin size, this will have a noticable effect on the quality of your models, see <a href="/ChromOptimise/ChromOptimise/Factors-that-affect-the-output#bin-size">this wiki page</a> for further information.</p>
<p><strong>Note</strong>: This file has commented lines at the end. Uncomment these lines to make the script remove binary files associated with mitochondrial DNA.</p>
<p>Example:</p>
<div class="language-shell codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-shell codeBlock_bY9V thin-scrollbar" style="color:#393A34;background-color:#f6f8fa"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># Takes all files in ${SUBSAMPLE_DIR} that use a sample size of 75% and</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># binarizes them using a bin size of 200 and the hg19 assembly</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sbatch 4_BinarizeFiles.sh 200 75 hg19</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div>
<h2 class="anchor anchorWithStickyNavbar_LWe7" id="5_batch_createincrementalmodelssh">5_batch_CreateIncrementalModels.sh<a href="#5_batch_createincrementalmodelssh" class="hash-link" aria-label="Direct link to 5_batch_CreateIncrementalModels.sh" title="Direct link to 5_batch_CreateIncrementalModels.sh">​</a></h2>
<p>This uses ChromHMM&#x27;s <code>LearnModel</code> command to learn multiple hidden Markov models. It also obtains the <a href="/ChromOptimise/ChromOptimise/ChromHMM-overview#estimated-log-likelihood">estimated log likelihood</a> value of each model.</p>
<p><strong>Note</strong>: This script uses the &#x27;information&#x27; initialisation method for the starting parameter set for the model. Due to this, the number of states in the model cannot exceed the total number of combinations of marks in your dataset.
<br>
<!-- -->One may not know this number before running the script. However, the maximum number of mark combinations one can have for a binary file will always be less than or equal to <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msup><mn>2</mn><mi>k</mi></msup></mrow><annotation encoding="application/x-tex">2^k</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.8491em"></span><span class="mord"><span class="mord">2</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8491em"><span style="top:-3.063em;margin-right:0.05em"><span class="pstrut" style="height:2.7em"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight" style="margin-right:0.03148em">k</span></span></span></span></span></span></span></span></span></span></span>, where <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>k</mi></mrow><annotation encoding="application/x-tex">k</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.6944em"></span><span class="mord mathnormal" style="margin-right:0.03148em">k</span></span></span></span> is the total number of marks. If one exceeds the maximum number of states permitted for this initialisation technique, ChromHMM will tell the user the maximum number of states allowed in the associated error message.</p>
<p>This script will always generate a model with 2 states. Inspecting the emission parameters for this simple model is a good way of validating your data.
<br>
<!-- -->Generally, most of your genomic data will be non-coding regions. The 2 state model should reflect this property in the overlap files produced.</p>
<p>Example:</p>
<div class="language-shell codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-shell codeBlock_bY9V thin-scrollbar" style="color:#393A34;background-color:#f6f8fa"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># Generates 6 models using the hg19 assembly (sizes 2-7) using </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># the binarized files with bin size 200 and sample size 75%</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># --array=1-3 is used so that each array element learns 2 models (optimal).</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sbatch --array=1-3 5_batch_CreateIncementalModels.sh 6 1 200 75 hg19</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div>
<h2 class="anchor anchorWithStickyNavbar_LWe7" id="6_optimalnumberofstatessh">6_OptimalNumberOfStates.sh<a href="#6_optimalnumberofstatessh" class="hash-link" aria-label="Direct link to 6_OptimalNumberOfStates.sh" title="Direct link to 6_OptimalNumberOfStates.sh">​</a></h2>
<p>This is the statistical analysis step where the optimal number of states to use with the model is determined (subject to many <a href="/ChromOptimise/ChromOptimise/Factors-that-affect-the-output">factors</a>).</p>
<p>This script uses four R scripts:</p>
<ul>
<li>PlotLikelihoods.R<!-- -->
<ul>
<li>This creates a plot of estimated log likelihood against number of model states. This gives the user feedback on how the accuracy of the model increases as one increases the number of states used in the model.</li>
</ul>
</li>
<li>OptimumNumberOfStates.R<!-- -->
<ul>
<li>This uses vector similarity (Euclidean distance) to identify how similar each pair states are in terms of their emission parameters in each model.</li>
<li>The script also looks at the maximum transition probability towards each state to determine the relative assignment of each state (a low maximum transition probability implies that the <a href="https://github.com/sof202/ChromOptimise/wiki/ChromHMM-overview#estimated-log-likelihood" target="_blank" rel="noopener noreferrer">forwards-backwards algorithm</a> assigns the state to relatively few genomic bins).<!-- -->
<ul>
<li>If a state scores low enough in both of these metrics (threshold is defined in <a href="/ChromOptimise/ChromOptimise/Configuration-Files-Setup#configr">config.R</a>), then the state is classed as redundant.</li>
</ul>
</li>
<li>Models with redundant states are rejected in favour of less complex models.</li>
</ul>
</li>
<li>CalculateAIC.R<!-- -->
<ul>
<li>This script will calculate the <a href="https://en.wikipedia.org/wiki/Akaike_information_criterion" target="_blank" rel="noopener noreferrer">Akaike information critereon</a> for each of the models created in <a href="#5_batch_createincrementalmodelssh">5_batch_CreateIncremenatlModels.sh</a>.</li>
<li>AIC is a heuristic for probabilistic models that attempts to explain how accurate a model is in spite of its complexity.<!-- -->
<ul>
<li>It is given by: <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>A</mi><mi>I</mi><mi>C</mi><mo>=</mo><mn>2</mn><mi>k</mi><mo>−</mo><mn>2</mn><mi>l</mi><mi>n</mi><mo stretchy="false">(</mo><mi>L</mi><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">AIC=2k-2ln(L)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.6833em"></span><span class="mord mathnormal">A</span><span class="mord mathnormal" style="margin-right:0.07847em">I</span><span class="mord mathnormal" style="margin-right:0.07153em">C</span><span class="mspace" style="margin-right:0.2778em"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2778em"></span></span><span class="base"><span class="strut" style="height:0.7778em;vertical-align:-0.0833em"></span><span class="mord">2</span><span class="mord mathnormal" style="margin-right:0.03148em">k</span><span class="mspace" style="margin-right:0.2222em"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222em"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em"></span><span class="mord">2</span><span class="mord mathnormal" style="margin-right:0.01968em">l</span><span class="mord mathnormal">n</span><span class="mopen">(</span><span class="mord mathnormal">L</span><span class="mclose">)</span></span></span></span>.<!-- -->
<ul>
<li>Where <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>k</mi></mrow><annotation encoding="application/x-tex">k</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.6944em"></span><span class="mord mathnormal" style="margin-right:0.03148em">k</span></span></span></span> is the number of parameters in the model and <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>L</mi></mrow><annotation encoding="application/x-tex">L</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.6833em"></span><span class="mord mathnormal">L</span></span></span></span> is the estimated likelihood.</li>
</ul>
</li>
<li>If two models have the same log likelihood, but one has more states, the model with fewer states will have a lower (and therefore better) AIC.</li>
</ul>
</li>
<li>Considering the estimated log likelihood is generally very low for ChromHMM, these values are very large. To combat this, we employ the relative AIC measure: <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>A</mi><mi>I</mi><msub><mi>C</mi><mrow><mi>r</mi><mi>e</mi><mi>l</mi><mi>a</mi><mi>t</mi><mi>i</mi><mi>v</mi><mi>e</mi></mrow></msub><mo>=</mo><mi>e</mi><mi>x</mi><mi>p</mi><mo stretchy="false">(</mo><mo stretchy="false">(</mo><mi>A</mi><mi>I</mi><msub><mi>C</mi><mrow><mi>m</mi><mi>i</mi><mi>n</mi></mrow></msub><mo>−</mo><mi>A</mi><mi>I</mi><msub><mi>C</mi><mi>i</mi></msub><mo stretchy="false">)</mo><mi mathvariant="normal">/</mi><mn>2</mn><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">AIC_{relative} = exp((AIC_{min}-AIC_i)/2)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.8333em;vertical-align:-0.15em"></span><span class="mord mathnormal">A</span><span class="mord mathnormal" style="margin-right:0.07847em">I</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.07153em">C</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361em"><span style="top:-2.55em;margin-left:-0.0715em;margin-right:0.05em"><span class="pstrut" style="height:2.7em"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">re</span><span class="mord mathnormal mtight" style="margin-right:0.01968em">l</span><span class="mord mathnormal mtight">a</span><span class="mord mathnormal mtight">t</span><span class="mord mathnormal mtight">i</span><span class="mord mathnormal mtight" style="margin-right:0.03588em">v</span><span class="mord mathnormal mtight">e</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2778em"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2778em"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em"></span><span class="mord mathnormal">e</span><span class="mord mathnormal">x</span><span class="mord mathnormal">p</span><span class="mopen">((</span><span class="mord mathnormal">A</span><span class="mord mathnormal" style="margin-right:0.07847em">I</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.07153em">C</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3117em"><span style="top:-2.55em;margin-left:-0.0715em;margin-right:0.05em"><span class="pstrut" style="height:2.7em"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">min</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222em"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222em"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em"></span><span class="mord mathnormal">A</span><span class="mord mathnormal" style="margin-right:0.07847em">I</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.07153em">C</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3117em"><span style="top:-2.55em;margin-left:-0.0715em;margin-right:0.05em"><span class="pstrut" style="height:2.7em"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em"><span></span></span></span></span></span></span><span class="mclose">)</span><span class="mord">/2</span><span class="mclose">)</span></span></span></span>. This measure is proportional to the probability that the <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msup><mi>i</mi><mrow><mi>t</mi><mi>h</mi></mrow></msup></mrow><annotation encoding="application/x-tex">i^{th}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.8491em"></span><span class="mord"><span class="mord mathnormal">i</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8491em"><span style="top:-3.063em;margin-right:0.05em"><span class="pstrut" style="height:2.7em"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mathnormal mtight">t</span><span class="mord mathnormal mtight">h</span></span></span></span></span></span></span></span></span></span></span></span> model minimizes the (estimated) information loss (<a href="https://link.springer.com/book/10.1007/b97636" target="_blank" rel="noopener noreferrer">source</a>).</li>
</ul>
</li>
<li>CalculateBIC.R<!-- -->
<ul>
<li>The Bayesian information critereon is very similar to AIC. The difference here is that this is much more punishing on the number of parameters in the model when the dataset is very large (which it will be in the case of genomic datasets).</li>
<li>The Bayesian information critereon is given by: <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>B</mi><mi>I</mi><mi>C</mi><mo>=</mo><mi>l</mi><mi>n</mi><mo stretchy="false">(</mo><mi>n</mi><mo stretchy="false">)</mo><mi>k</mi><mo>−</mo><mn>2</mn><mi>l</mi><mi>n</mi><mo stretchy="false">(</mo><mi>L</mi><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">BIC = ln(n)k - 2ln(L)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.6833em"></span><span class="mord mathnormal" style="margin-right:0.05017em">B</span><span class="mord mathnormal" style="margin-right:0.07847em">I</span><span class="mord mathnormal" style="margin-right:0.07153em">C</span><span class="mspace" style="margin-right:0.2778em"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2778em"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em"></span><span class="mord mathnormal" style="margin-right:0.01968em">l</span><span class="mord mathnormal">n</span><span class="mopen">(</span><span class="mord mathnormal">n</span><span class="mclose">)</span><span class="mord mathnormal" style="margin-right:0.03148em">k</span><span class="mspace" style="margin-right:0.2222em"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222em"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em"></span><span class="mord">2</span><span class="mord mathnormal" style="margin-right:0.01968em">l</span><span class="mord mathnormal">n</span><span class="mopen">(</span><span class="mord mathnormal">L</span><span class="mclose">)</span></span></span></span>
<ul>
<li>Where <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>n</mi></mrow><annotation encoding="application/x-tex">n</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.4306em"></span><span class="mord mathnormal">n</span></span></span></span> is the total number of observations, <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>k</mi></mrow><annotation encoding="application/x-tex">k</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.6944em"></span><span class="mord mathnormal" style="margin-right:0.03148em">k</span></span></span></span> is the number of parameters in the model and <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>L</mi></mrow><annotation encoding="application/x-tex">L</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.6833em"></span><span class="mord mathnormal">L</span></span></span></span> is the estimated likelihood.</li>
</ul>
</li>
<li>Again, these values are very large and hard to compare. As a result we find the BIC of each model relative to the minimum BIC of all models (this time using a simple ratio as there is no merit to using <span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>e</mi><mi>x</mi><mi>p</mi></mrow><annotation encoding="application/x-tex">exp</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.625em;vertical-align:-0.1944em"></span><span class="mord mathnormal">e</span><span class="mord mathnormal">x</span><span class="mord mathnormal">p</span></span></span></span> with this heuristic).<!-- -->
<ul>
<li>As with AIC, a smaller value for BIC generally indicates a better model.</li>
</ul>
</li>
<li>It is important to note that AIC and BIC are heuristics, not metrics. They only approximate how good a model is relative to its complexity. As such the graphics and .csv files that come from these scripts are in place to provide further information to the user (Perhaps a model with 6 states is better under this metric than a model with 8 states, despite the more complex model having no &#x27;redundant states&#x27;).</li>
</ul>
</li>
</ul>
<p>Example:</p>
<div class="language-shell codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-shell codeBlock_bY9V thin-scrollbar" style="color:#393A34;background-color:#f6f8fa"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># Generates metrics and heuristics for the models currently in ${MODEL_DIR}</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># and uses these to determine  the optimal number of states for the binarized</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># data set.</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sbatch 6_OptimalNumberOfStates.sh</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div>
<h2 class="anchor anchorWithStickyNavbar_LWe7" id="note">Note<a href="#note" class="hash-link" aria-label="Direct link to Note" title="Direct link to Note">​</a></h2>
<p>This pipeline was built using the SLURM workload manager. Scripts will result in errors or even fail entirely if not ran using the <code>sbatch</code> SLURM command. Please adapt the scripts accordingly if SLURM is not being used. For information on how to do this, please consult <a href="/ChromOptimise/ChromOptimise/SLURM-Workload-Manager-Information">the page on SLURM</a>.</p></div></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages"><a class="pagination-nav__link pagination-nav__link--prev" href="/ChromOptimise/ChromOptimise/Configuration-Files-Setup"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">Configuration files setup</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/ChromOptimise/ChromOptimise/Supplementary-pipeline-explanation"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Supplementary pipeline explanation</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_bqdL thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#0_egadownloadingsh" class="table-of-contents__link toc-highlight">0_EGADownloading.sh</a></li><li><a href="#1_movefilestosingledirectorysh" class="table-of-contents__link toc-highlight">1_MoveFilesToSingleDirectory.sh</a></li><li><a href="#2_batch_processbamfilessh" class="table-of-contents__link toc-highlight">2_batch_ProcessBamFiles.sh</a></li><li><a href="#3_subsamplebamfilessh" class="table-of-contents__link toc-highlight">3_SubsampleBamFiles.sh</a></li><li><a href="#4_binarizefilessh" class="table-of-contents__link toc-highlight">4_BinarizeFiles.sh</a></li><li><a href="#5_batch_createincrementalmodelssh" class="table-of-contents__link toc-highlight">5_batch_CreateIncrementalModels.sh</a></li><li><a href="#6_optimalnumberofstatessh" class="table-of-contents__link toc-highlight">6_OptimalNumberOfStates.sh</a></li><li><a href="#note" class="table-of-contents__link toc-highlight">Note</a></li></ul></div></div></div></div></main></div></div></div><footer class="footer footer--dark"><div class="container container-fluid"><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2024 ChromOptimise, Inc. Built with Docusaurus.</div></div></div></footer></div>
</body>
</html>